### PR TITLE
feat (invoices): handle payment retry for specific invoice

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

Add API and graphQL actions for triggering specific invoice payment retry

## Description

When invoice payment status is pending or failed we want to be able to retry payment. It's allowed to trigger such an action from both graphQL and API.

### CODE REVIEW

File: app/services/invoices/payments/gocardless_service.rb [hunk 1]
```diff
@@ -56,7 +58,10 @@ def update_payment_status(provider_payment_id:, status:)
         payment.update!(status: status)
 
         invoice_payment_status = invoice_payment_status(status)
-        payment.invoice.update!(payment_status: invoice_payment_status)
+        payment.invoice.update!(
+          payment_status: invoice_payment_status,
+          ready_for_payment_processing: invoice_payment_status != 'succeeded'
```
Comment (jdenquin) [inline app/services/invoices/payments/gocardless_service.rb:63]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/invoices/payments/retry_service.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      def call
+        return result.not_found_failure!(resource: 'invoice') if invoice.blank?
+        # TODO: Handle draft invoice when grace period feature is merged
+        return result.not_allowed_failure!(code: 'invalid_status') if invoice.succeeded?
+
+        unless invoice.ready_for_payment_processing?
+          return result.not_allowed_failure!(code: 'payment_processor_is_currently_handling_payment')
+        end
+
+        webhook_type = invoice.invoice_type == 'subscription' ? :invoice : invoice.invoice_type.to_sym
```
Comment (jdenquin) [inline app/services/invoices/payments/retry_service.rb:21]: Style/TernaryParentheses: Use parentheses for ternary expressions with complex conditions.

File: db/migrate/20221216154033_add_payment_retry_columns_to_invoices.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class AddPaymentRetryColumnsToInvoices < ActiveRecord::Migration[7.0]
+  def change
+    add_column :invoices, :payment_attempts, :integer, default: 0, null: false
```
Comment (jdenquin) [inline db/migrate/20221216154033_add_payment_retry_columns_to_invoices.rb:5]: Rails/BulkChangeTable: You can use `change_table :invoices, bulk: true` to combine alter queries.

File: app/services/invoices/payments/gocardless_service.rb [hunk 4]
```diff
@@ -56,7 +58,10 @@ def update_payment_status(provider_payment_id:, status:)
         payment.update!(status: status)
 
         invoice_payment_status = invoice_payment_status(status)
-        payment.invoice.update!(payment_status: invoice_payment_status)
+        payment.invoice.update!(
+          payment_status: invoice_payment_status,
+          ready_for_payment_processing: invoice_payment_status != 'succeeded',
```
Comment (rsempe) [inline app/services/invoices/payments/gocardless_service.rb:63]: `failed` is also considered as ready_for_payment_processing, then?

File: app/services/invoices/payments/gocardless_service.rb [hunk 5]
```diff
@@ -121,7 +126,7 @@ def create_gocardless_payment
             },
           },
           headers: {
-            'Idempotency-Key' => invoice.id,
+            'Idempotency-Key' => "#{invoice.id}-#{invoice.payment_attempts}",
```
Comment (rsempe) [inline app/services/invoices/payments/gocardless_service.rb:129]:
```suggestion
            'Idempotency-Key' => "#{invoice.id}/#{invoice.payment_attempts}",
```

File: app/services/invoices/payments/stripe_service.rb [hunk 6]
```diff
@@ -105,7 +107,7 @@ def create_stripe_payment
           stripe_payment_payload,
           {
             api_key: stripe_api_key,
-            idempotency_key: invoice.id,
+            idempotency_key: "#{invoice.id}-#{invoice.payment_attempts}",
```
Comment (rsempe) [inline app/services/invoices/payments/stripe_service.rb:110]:
```suggestion
            idempotency_key: "#{invoice.id}/#{invoice.payment_attempts}",
```

**Review by rsempe — APPROVED** [review]

File: app/services/invoices/payments/gocardless_service.rb (hunk 4 above)
Comment (lovrocolic) [reply app/services/invoices/payments/gocardless_service.rb:63]: yep:)

File: app/services/invoices/payments/stripe_service.rb (hunk 6 above)
Comment (lovrocolic) [reply app/services/invoices/payments/stripe_service.rb]: fixed

File: app/services/invoices/payments/gocardless_service.rb (hunk 5 above)
Comment (lovrocolic) [reply app/services/invoices/payments/gocardless_service.rb]: fixed

**Pull request merged**